File: src/main/java/com/github/happylynx/prick/lib/commands/IndexToTree.java
package com.github.happylynx.prick.lib.commands;

import com.github.happylynx.prick.lib.model.*;

import java.nio.file.Path;
import java.util.*;
import java.util.stream.Collectors;

public class IndexToTree {

    private final Stack<TreeItem> stack = new Stack<>();

    private IndexToTree() {}

    public static HashId convert(Index index) {
        final IndexToTree indexToTree = new IndexToTree();
        index.getItems().stream().forEachOrdered(indexToTree::nextItem);
        indexToTree.finish();
        return indexToTree.getResult();
    }

    private void finish() {
        while (stack.size() > 1) {
            final List<TreeItem> dirItems = extractLastDirItems();
            final TreeHash treeHash = computeNStoreTree(dirItems);
            stack.push(treeHash);
        }
    }

    private HashId getResult() {
        if (stack.isEmpty()) {
            return ObjectStorage.store("", pri);
        }
        return stack.pop().getHash();
    }

    private void nextItem(IndexItem item) {
        stack.push(item);
        while (directoryFinished()) {
            storeDirectory();
        }
    }

    private void storeDirectory() {
        final TreeItem lastItem = stack.pop();
        final List<TreeItem> completedDirItems = extractLastDirItems();
        final TreeHash tree = computeNStoreTree(completedDirItems);
        stack.push(tree);
        stack.push(lastItem);
    }

    private TreeHash computeNStoreTree(List<TreeItem> items) {
        final String treeFileContent = items.stream()
                .map(FileFormats::createTreeLine)
                .collect(Collectors.joining(FileFormats.LINE_SEPARATOR));
        final HashId treeHash = ObjectStorage.store(treeFileContent);
        return new TreeHash(treeHash, items.get(0).getPath().getParent());
    }

    private List<TreeItem> extractLastDirItems() {
        final ArrayList<TreeItem> result = new ArrayList<>();
        final Path parentDir = stack.peek().getPath().getParent();
        while (stack.peek().getPath().getParent().equals(parentDir)) {
            result.add(stack.pop());
        }
        Collections.reverse(result);
        return result;
    }

    private boolean directoryFinished() {
        if (stack.size() < 2) {
            return false;
        }
        final Path lastDir = stack.get(stack.size() - 1).getPath().getParent();
        final Path previousDir = stack.get(stack.size() - 2).getPath().getParent();
        return !lastDir.equals(previousDir) && !lastDir.startsWith(previousDir);

    }
}

File: src/main/kotlin/com/github/happylynx/prick/lib/model/model2/IndexToTree.kt
package com.github.happylynx.prick.lib.model.model2

import com.github.happylynx.prick.lib.commands.FileFormats
import com.github.happylynx.prick.lib.commands.ObjectStorage
import com.github.happylynx.prick.lib.commands.PrickContext
import com.github.happylynx.prick.lib.model.HashId
import com.github.happylynx.prick.lib.model.Index
import java.lang.IllegalStateException
import java.lang.IndexOutOfBoundsException
import java.nio.file.Path
import java.util.*

class IndexToTree(private val ctx: PrickContext) {

    companion object {
        fun run(index: Index, ctx: PrickContext): HashId {
            return IndexToTree(ctx).process(index)
        }
    }

    private val stack: Stack<IndexToTreeStackItem> = Stack()

    private fun process(index: Index): HashId {
        AddDirectoryBounds.add(index.items.values.stream()).forEach(::processItem)
        return finish()
    }

    private fun finish(): HashId {
        return (stack.peek() as TreeItem.TreeDirectory).hash
    }

    private fun processItem(item: IndexStreamItem) {
        val itemToPush: IndexToTreeStackItem = when (item) {
            is IndexStreamItem.DirBegin,
            is IndexStreamItem.DirEnd -> item as IndexToTreeStackItem
            is IndexItem.FileIndexItem -> item.toTreeItem()
            is IndexItem.SymlinkIndexItem -> item.toTreeItem()
            else -> throw IllegalStateException()
        }
        stack.push(itemToPush)
        if (item is IndexStreamItem.DirEnd) {
            wrapDirectory()
        }
    }

    private fun wrapDirectory() {
        val children = mutableSetOf<TreeItem>()
        var dirPath = (stack.pop() as IndexStreamItem.DirEnd).path
        while (stack.peek() !is IndexStreamItem.DirBegin) {
            children.add(stack.pop() as TreeItem)
        }
        var dirBegin = stack.pop() as IndexStreamItem.DirBegin
        assert(dirBegin.path == dirPath)
        val dir: TreeItem.TreeDirectory = storeDirectory(dirPath.fileName.toString(), children)
        stack.push(dir)
    }

    private fun storeDirectory(name: String, items: Set<TreeItem>): TreeItem.TreeDirectory{
        val data = FileFormats.Tree.serialize(items)
        val hash = ObjectStorage.store(data, ctx)
        return TreeItem.TreeDirectory(name, hash)
    }

    interface IndexToTreeStackItem
}